### ISSUE TITLE
Reference-counted types to enable flexible links between objects

### ISSUE THREAD

**Issue by hoijnet** [terminusdb/terminusdb#2111]
**Is your feature request related to a problem? Please describe.**
Today one can either create a document type to constrain RDF triples into a shared lifecycle, or create a subdocument to add a "layer" of information to a type (JSON-LD). 

An elegant solution for bi-directional links would be to have a reference-counted linked relationship between two objects. Imagine the concept of friendship, being between two persons, seeing it as directional gives a skewed view of reality, a bi-directional link as a reference-counted third object that connects the individiual would be quite handy.

Another similar instance is a configuration of items into an offering. It could of course be modelled as a type with a set of links, either as individual properties, or as a set of link with a Cardinality. However, if any of the parts of the offering is removed, the offering could be relevant to automatically delete as well (as a design choice).

**Describe the solution you'd like**
Having reference counted type properties (if one of the property linked is removed, the entire object should be removed too) would enable a significant opportunity to avoid having to write custom logic and instead have more operations possible to perform over the REST and GraphQL interface.

**Describe alternatives you've considered**
WOQL logic for it, but would be hard for the use case.

**Additional context**
N/A

**Comment by dfrnt-SigmundFuchs** [comment]
This issue has been closed due to inactivity. Please don’t hesitate to reopen it if you have any further questions or would like to continue the discussion.

**Comment by hoijnet** [comment]
A more elegant solution to this was proposed by @timathom here: https://github.com/orgs/terminusdb/discussions/1978

In essence, there are two ways of making this happen, one is some kind of metadata annotation on which fields should be reference counted and a custom post-commit action. The other is to implement `@unfoldable` and `@subdocument` (find a better metadata name for it) as metadata on the linked field to indicate that the linked document is to be treated as a (possibly shared) subdocument and garbage collect the linked document when it is no longer referenced by any documents.

This would enable bi-directional links to be easily created (this issue), and also solve the challenge for @timathom.

**Comment by Tails** [comment]
This sounds a lot like CASCADE DELETE from the SQL paradigm, which would indeed very much come in handy. 

So just to be clear, this is _not_ about having a reference field populated on a target model when it is assigned on a source reference field right? I often have relations modeled on both sides (the bidirectional part), but often at insertion/creation time only one side is available in the application-layer context. Would also be nice to have that 'auto-bi-directionality' so that a link is immediately set on a reference target backward as well!

**Comment by hoijnet** [comment]
> reference an instance of a class as _either_ a top-level document or nested child/subdocument

@timathom Taking another look at this. That's the idea. Thanks for the feedback. It took a while to tackle this one.

The semantics will be document. Main idea looks to be that `@child` annotations on a property means that a deletion will be attempted, and if no other @child annotations are left and there are no mandatory multiplicity predicates, then the shared document will be removed (even if it is part of a set or optional somewhere).

For the `@shared` classes on the other hand, if there are any references to them left, no matter the type, they will be kept. Where there are no links to it anymore, they will be deleted. Special case though, it will be possible to create classes with `@shared` annotation without links, but one must be careful, because if they are later linked to another object, and then the original object is deleted and no other documents link to them, the formerly free-floating document will be deleted.

The alternative would be to not allow @shared documents to exist standalone, which I think does not have to be enforced as it removes flexibility to create objects that are later joined into a bigger organism.

Comments are more than welcome!

**Comment by hoijnet** [comment]
@timathom @Tails the `@shared` child documents is now a reality, complementing `@subdocuments` for different tasks. Initial testing yields improvement performance over `@subdocument` documents as well. Documentation is being prepared as well (pasting the main document below). See #2437 for more details. @timathom, can you check if it covers your specific needs and can provide feedback?

We see numerous use cases solved by this and decided not to implement `@child` annotations on fields as they could be hard to reason about, and that is was hard to come up with clear semantics. For `@shared`, it's clear.

See draft documentation below.

---

## Document types at a glance

TerminusDB offers several RDF graph document types, each with different ownership, lifecycle, and identity semantics. Choosing the right type shapes how your linked data behaves when documents are created, shared between parents, and deleted.

This page compares the types side by side to help you choose the right one for your use case.

## Comparison table

| Property | Regular document | Shared document (`@shared`) | Subdocument (`@subdocument`) |
|----------|-----------------|----------------------------|------------------------------|
| **Identity** | Own IRI (`Class/id`) | Own IRI (`Class/id`) | Nested IRI (parent path) |
| **Key types** | Any (`Lexical`, `Hash`, `ValueHash`, `Random`) | Any (`Lexical`, `Hash`, `ValueHash`, `Random`) | Any (`Lexical`, `Hash`, `ValueHash`, `Random`) |
| **Ownership** | Independent | Independent (but lifecycle-managed) | Owned by exactly one parent document/subdocument |
| **Can be referenced by** | Any number of documents | Any number of documents | Only its containing document |
| **Lifecycle** | Manual deletion only | Automatic cascade deletion when zero references remain, or manually | Only deleted when parent is deleted |
| **Standalone creation** | Yes | Yes | No (must be nested in a parent) |
| **Standalone retrieval** | Yes (by IRI) | Yes (by IRI) | Yes (by its nested IRI) |
| **Unfolding** | By IRI reference (unless `@unfoldable`) | By IRI reference (unless `@unfoldable`) | Always unfolded inline |

## Regular documents

Regular documents are the default. They have their own IRI, can be referenced/linked by any number of other documents, and are never deleted automatically. Use them when documents have an independent lifecycle.

```json
{
    "@type"  : "Class",
    "@id"    : "Person",
    "@key"   : { "@type": "Lexical", "@fields": ["name"] },
    "name"   : "xsd:string",
    "email"  : "xsd:string"
}
```

**When to use:** The document exists independently, is shared across many contexts, and should never be automatically deleted. Examples: users, products, reference data.

## Subdocuments (`@subdocument`)

A subdocument is wholly owned by its containing document. It cannot exist independently and is always created, retrieved, and deleted together with its parent.

```json
{
    "@type"         : "Class",
    "@id"           : "Address",
    "@subdocument"  : [],
    "@key"          : { "@type": "Random" },
    "street"        : "xsd:string",
    "city"          : "xsd:string",
    "postal_code"   : "xsd:string"
}
```

**When to use:** The data is internal to a parent document and never needs to be shared or referenced from elsewhere. Examples: addresses embedded in a person, configuration nested in a service, line items in an order. All documents that belong only to one particular parent document. 

**Constraints:**
- Only one parent can point to a given subdocument instance
- Cannot be retrieved or updated independently (only through the parent)

## Shared documents (`@shared`)

A shared document is a regular document (own IRI, any key type) that participates in automatic lifecycle management. When a `@shared` document is no longer referenced by any other document, it is automatically cascade-deleted at commit time.

```json
{
    "@type"    : "Class",
    "@id"      : "Footnote",
    "@shared"  : [],
    "@key"     : { "@type": "Lexical", "@fields": ["label"] },
    "label"    : "xsd:string",
    "content"  : "xsd:string"
}
```

**When to use:** The document may be shared across multiple parents, but once nothing references it, it should be automatically cleaned up. Examples: footnotes shared between articles, tags used by multiple resources, shared configuration blocks.

**Constraints:**
- Cannot also be `@subdocument` (mutual exclusion)
- Liveness is determined at commit time by counting all triples pointing to the document
- Explicit deletion is still permitted (behaves like a regular document delete)

### How `@shared` liveness works

When any reference to a `@shared` document is removed (through parent deletion, field update, or triple removal), the engine checks at commit time whether any triples remain that point to the target. If zero references remain, the document is automatically cascade-deleted.

The liveness check counts **all** triples pointing to the target document, regardless of which field or class they originate from. Any reference from any document keeps the `@shared` document alive.

```
Article A ──notes──► Footnote/1 (@shared)
Article B ──notes──► Footnote/1

Delete Article A → Footnote/1 still alive (Article B references it)
Delete Article B → Footnote/1 has zero references → cascade-deleted
```

### Cascade is recursive

If a cascade-deleted `@shared` document itself has fields pointing to other `@shared` documents, those targets are checked in turn. A visited set prevents infinite loops on circular references.

If two `@shared` documents reference only each other (a circular island) and all external references are removed, both are cascade-deleted.

### Standalone creation

`@shared` documents can be created independently with no parent referencing them. They are not immediately deleted upon creation; the liveness check only fires when a reference is **removed**, not when a document has never been referenced.

### Worked example: full lifecycle

Using the Footnote/Article schema from above, this example walks through creation, sharing, and cascade deletion.

**Step 1 — Define the schema:**

```json
{ "@type": "Class", "@id": "Footnote", "@shared": [],
  "@key": { "@type": "Lexical", "@fields": ["label"] },
  "label": "xsd:string", "content": "xsd:string" }

{ "@type": "Class", "@id": "Article",
  "@key": { "@type": "Lexical", "@fields": ["title"] },
  "title": "xsd:string",
  "notes": { "@type": "Set", "@class": "Footnote" } }
```

**Step 2 — Create a `@shared` footnote standalone (no parent yet):**

```json
{ "@type": "Footnote", "label": "fn1", "content": "See appendix A." }
```

Result: `Footnote/fn1` exists. No cascade check fires — it has never been referenced.

**Step 3 — Insert two articles that reference the footnote:**

```json
{ "@type": "Article", "title": "intro", "notes": ["Footnote/fn1"] }
{ "@type": "Article", "title": "appendix", "notes": ["Footnote/fn1"] }
```

Result: `Footnote/fn1` is now referenced by two parents.

**Step 4 — Delete one parent:**

```json
DELETE Article/intro
```

Result: `Article/intro` is deleted. The engine checks `Footnote/fn1` — `Article/appendix` still references it. **Target survives.**

**Step 5 — Delete the second parent:**

```json
DELETE Article/appendix
```

Result: `Article/appendix` is deleted. The engine checks `Footnote/fn1` — zero references remain. **Target is cascade-deleted automatically.**

After step 5, querying for `Footnote/fn1` returns nothing — it was cleaned up without an explicit delete call.

## Unfoldable documents (`@unfoldable` / `@unfold`)

Unfolding is orthogonal to lifecycle and ownership. It controls how a document appears **on retrieval** - whether it is returned as a full inline object or as an IRI reference.

| Annotation | Level | Purpose |
|-----------|-------|---------|
| `@unfoldable` | Class-level | All references to this class unfold inline |
| `@unfold` | Field-level | Only this specific field unfolds its target inline |

Unfolding can be combined with any document type. A `@shared` document may also be `@unfoldable` if you want it to appear inline when retrieved. The annotations address different concerns:

- **Lifecycle:** `@shared`, `@subdocument` (when is the document deleted?)
- **Retrieval:** `@unfoldable`, `@unfold` (how is the document represented in API responses?)

See the [Document Unfolding Reference](/docs/document-unfolding-reference/) for details on cycle detection and performance.

## Decision flowchart

Use the following questions to choose the right document type:

1. **Does this data belong exclusively to one parent?**
   - Yes → **Subdocument** (`@subdocument`)
   - No → Continue

2. **Should multiple documents reference this data?**
   - No → **Regular document** (or subdocument if exclusively owned)
   - Yes → Continue

3. **Should the document be automatically deleted when nothing references it?**
   - Yes → **Shared document** (`@shared`)
   - No → **Regular document**

4. **Should the document appear inline when retrieved (regardless of type)?**
   - Yes → Add `@unfoldable` to the class, or `@unfold` to specific fields

## Migration notes

### Adding `@shared` to an existing class

No data migration is needed. Adding `@shared` to a class definition enables the cascade semantics from that point forward. Existing documents are unaffected until a reference to them is removed.

### Removing `@shared` from a class

The cascade semantics stop immediately. Previously-shared documents become regular documents with no automatic lifecycle management. No data changes are required.

### Moving from `@subdocument` to `@shared`

This is a breaking schema change. Subdocuments have nested IRIs and restricted key types; shared documents have independent IRIs. Migration requires:

1. Create new shared document instances with their own IRIs
2. Update all parent references to point to the new IRIs
3. Remove the old subdocument instances
4. Update the schema (remove `@subdocument`, add `@shared`)

This cannot currently be done in a single transaction without data restructuring.

**Comment by hoijnet** [comment]
There's still some work to do to verify edge cases, but for just testing the ideas on a branch, you can test the early version using this docker image:
* `docker.io/terminusdb/terminusdb-server:feature-shared-child-annotation`

**Comment by hoijnet** [comment]
And from the upcoming change to the schema reference guide:


### @shared

The `@shared` key is present with the value `[]` or it is not present.

A class designated as shared is a regular document (with its own IRI) that participates in automatic lifecycle management. Unlike a subdocument, a shared document can be referenced by multiple parents and supports any key type. Unlike a regular document, a shared document is automatically cascade-deleted when no other document references it. Unlike the subdocument, a shared document class must explicitly get an `@unfoldable` annotation, or `@unfold` on the field that references it, to automatically unfold as a subdocument does.

The `@shared` annotation is **mutually exclusive** with `@subdocument` — a class cannot be both. However, `@shared` may be combined with `@unfoldable` (lifecycle and retrieval are orthogonal concerns). Also, `@shared` is inherited and is not possible to override by `@unfoldable` (or the other way around). 

#### Liveness rule

When any reference to a `@shared` document is removed (through parent deletion, field update, or WOQL triple removal), the engine checks at commit time whether any triples remain that point to the target. The check counts **all** triples pointing to the target, regardless of which field or class they originate from. If zero references remain, the document is cascade-deleted.

Cascade is recursive: if the deleted document itself references other `@shared` documents, those targets are checked in turn. Circular islands (two or more `@shared` documents that reference only each other) are detected and deleted together.

#### Standalone creation

`@shared` documents can be created independently with no parent referencing them. The liveness check fires only when a reference is **removed**, not when a document has never been referenced.

#### Code: An example shared document declaration

```json
{
    "@type"        : "@context",
    "@base"        : "terminusdb://i/",
    "@schema"      : "terminusdb://s#"
}
{
    "@type"        : "Class",
    "@id"          : "Footnote",
    "@shared"      : [],
    "@key"         :
    {
        "@type"    : "Lexical",
        "@fields"  : ["label"]
    },
    "label"        : "xsd:string",
    "content"      : "xsd:string"
}
{
    "@type"        : "Class",
    "@id"          : "Article",
    "@key"         :
    {
        "@type"    : "Lexical",
        "@fields"  : ["title"]
    },
    "title"        : "xsd:string",
    "notes"        :
    {
        "@type"    : "Set",
        "@class"   : "Footnote"
    }
}
```

#### Code: Example shared document instances

```json
{
    "@type"        : "Footnote",
    "@id"          : "Footnote/fn1",
    "label"        : "fn1",
    "content"      : "See appendix A for derivation."
}
{
    "@type"        : "Article",
    "@id"          : "Article/intro",
    "title"        : "intro",
    "notes"        : ["Footnote/fn1"]
}
{
    "@type"        : "Article",
    "@id"          : "Article/appendix",
    "title"        : "appendix",
    "notes"        : ["Footnote/fn1"]
}
```

In this example, `Footnote/fn1` is shared between two articles. Deleting `Article/intro` leaves `Footnote/fn1` alive (still referenced by `Article/appendix`). Deleting `Article/appendix` afterwards removes the last reference, and `Footnote/fn1` is cascade-deleted automatically.

**Comment by Tails** [comment]
Awesome!!
